covquestions-editor-app/src/components/QuestionnaireExecution.tsx

```diff
@@ -66,61 +66,44 @@ export const QuestionnaireExecution: React.FC<QuestionnaireExecutionProps> = ({
   const [questionnaireEngine, setQuestionnaireEngine] = useState(new QuestionnaireEngine(currentQuestionnaire));
   const [currentQuestion, setCurrentQuestion] = useState<Question | undefined>(undefined);
   const [currentValue, setCurrentValue] = useState<Primitive | Array<Primitive> | undefined>(undefined);
-  const [result, setResult] = useState<Result[] | undefined>(undefined);
-  const [progress, setProgress] = useState<number>(0);
-  const [doRerender, setDoRerender] = useState(false);
 
   const classes = useStyles();
 
   function restartQuestionnaire() {
-    const engine = new QuestionnaireEngine(currentQuestionnaire);
-    const nextQuestion = engine.nextQuestion();
-
-    setResult(undefined);
-    setProgress(0);
-    setCurrentValue(undefined);
-    setQuestionnaireEngine(engine);
-    setCurrentQuestion(nextQuestion);
-    setDoRerender(true);
+    setQuestionnaireEngine(new QuestionnaireEngine(currentQuestionnaire));
   }
 
   function handleNextClick() {
     questionnaireEngine.setAnswer(currentQuestion!.id, currentValue);
     setCurrentValue(undefined);
-
-    const nextQuestion = questionnaireEngine.nextQuestion();
-    setProgress(questionnaireEngine.getProgress());
-    if (nextQuestion) {
-      setCurrentQuestion(nextQuestion);
-    } else {
-      setCurrentQuestion(undefined);
-      setResult(questionnaireEngine.getResults());
-    }
+    setCurrentQuestion(questionnaireEngine.nextQuestion());
   }
 
   function handleBackClick() {
     const { question, answer } = questionnaireEngine.previousQuestion(currentQuestion!.id);
     setCurrentValue(answer);
-    setProgress(questionnaireEngine.getProgress());
-    if (question) {
-      setCurrentQuestion(question);
-    } else {
-      setCurrentQuestion(undefined);
-      setResult(questionnaireEngine.getResults());
-    }
+    setCurrentQuestion(question);
   }
 
-  useEffect(restartQuestionnaire, [currentQuestionnaire]);
+  useEffect(() => {
+    setQuestionnaireEngine((prevEngine) => {
+      const newEngine = new QuestionnaireEngine(currentQuestionnaire);
+      newEngine.setAnswersPersistence(prevEngine.getAnswersPersistence());
+      return newEngine;
+    });
+  }, [currentQuestionnaire]);
 
   useEffect(() => {
-    if (doRerender) {
-      setDoRerender(false);
-    }
-  }, [doRerender]);
+    setCurrentQuestion(questionnaireEngine.nextQuestion());
+    setCurrentValue(undefined);
```

Comment:
shouldn't it set the value to undefined or the previous value?

Reply:
Yeah, you are right. This line should just be gone to keep a potentially already set value. I will remove it.

Reply:
Well. I move it to the restart function. Because there we want a clean state.